tests/dm.go

```diff
@@ -0,0 +1,597 @@
+	if err != nil && !errors.IsNotFound(err) {
+		return fmt.Errorf("failed to delete namespace[%s]: %v", DMMySQLNamespace, err)
+	}
+	return nil
+}
+
+// CleanDMTiDB cleans the downstream TiDB cluster for DM E2E tests.
+func CleanDMTiDB(cli *versioned.Clientset, kubeCli kubernetes.Interface) error {
+	if err := cli.PingcapV1alpha1().TidbClusters(DMTiDBNamespace).Delete(DMTiDBName, nil); err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	if err := kubeCli.CoreV1().Namespaces().Delete(DMTiDBNamespace, nil); err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	return nil
+}
+
+// CreateDMSources creates upstream MySQL sources into DM cluster.
+func CreateDMSources(fw portforward.PortForward, ns, masterSvcName string) error {
+	apiPath := "/apis/v1alpha1/sources"
+
+	type Req struct {
+		Op     int      `json:"op"`
+		Config []string `json:"config"`
+	}
+	type Resp struct {
+		Result bool   `json:"result"`
+		Msg    string `json:"msg"`
+	}
+
+	var req = Req{
+		Op:     1,
+		Config: dmSources,
+	}
+	data, err := json.Marshal(req)
+	if err != nil {
+		return fmt.Errorf("failed to marshal source create request, %v, %v", req, err)
+	}
+
+	return wait.Poll(5*time.Second, 3*time.Minute, func() (bool, error) {
+		localHost, localPort, cancel, err := portforward.ForwardOnePort(
+			fw, ns, fmt.Sprintf("svc/%s", masterSvcName), dmMasterSvcPort)
+		if err != nil {
+			log.Logf("failed to forward dm-master svc: %v", err)
+			return false, nil
+		}
+		defer cancel()
+
+		body, err := httputil.DoBodyOK(
+			&http.Client{Transport: &http.Transport{}},
+			fmt.Sprintf("http://%s:%d%s", localHost, localPort, apiPath),
+			"PUT",
+			bytes.NewReader(data))
+		if err != nil {
+			log.Logf("failed to create DM sources: %v", err)
+			return false, nil
+		}
+		var resp Resp
+		if err = json.Unmarshal(body, &resp); err != nil {
+			log.Logf("failed to unmarshal DM source create response, %s: %v", string(body), err)
+			return false, nil
+		} else if !resp.Result && !strings.Contains(resp.Msg, "already exists") {
+			log.Logf("failed to create DM sources, %s: %v", resp.Msg, err)
+			return false, nil
+		}
+		return true, nil
+	})
+}
+
+// GenDMFullData generates full stage data for upstream MySQL.
+func GenDMFullData(fw portforward.PortForward, ns string) error {
+	var eg errgroup.Group
+	for i := int32(0); i < DMMySQLReplicas; i++ {
+		ordinal := i
+		eg.Go(func() error {
+			localHost, localPort, cancel, err := portforward.ForwardOnePort(
+				fw, DMMySQLNamespace, fmt.Sprintf("pod/%s-%d", DMMySQLSvcStsName, ordinal), uint16(DMMySQLPort))
+			if err != nil {
+				return fmt.Errorf("failed to forward MySQL[%d] pod: %v", ordinal, err)
+			}
+			defer cancel()
+
+			db, err := openDB(localHost, localPort)
+			if err != nil {
+				return err
+			}
+
+			// use ns as the database name.
+			// NOTE: we don't handle `already exists` or `duplicate entry` error now because we think ns is unqiue and the database instances are re-setup for each running.
```

Comment:
but we create `DMMySQLReplicas` time database with the same `ns` here, it will fail when DMMySQLReplicas is greater than 1 ?

Reply:
every `It` should have different `ns`. and in one `ns`, each replica should create database once.